File: backend/app/embeddings.py
"""Real semantic embeddings — same Perplexity endpoint/model as the Next.js side
(lib/embeddings.ts), so embeddings written from either stack are directly
comparable. Perplexity's embeddings endpoint requires its own model names and
rejects the standard OpenAI "float" encoding — verified directly against the
live API; only `base64_int8` and `base64_binary` are accepted.
"""

import base64
import struct

import httpx

from app.config import settings

EMBED_MODEL = "pplx-embed-v1-0.6b"
EMBED_URL = "https://api.perplexity.ai/v1/embeddings"


def _decode_int8(b64: str) -> list[int]:
    raw = base64.b64decode(b64)
    return list(struct.unpack(f"{len(raw)}b", raw))


def embed_text_sync(text: str) -> list[int] | None:
    """Embeds a single text. Returns None (rather than raising) if no API key is
    configured, so entity extraction never breaks just because embeddings aren't
    set up — callers should treat a missing embedding as "not yet searchable",
    not as an error.
    """
    if not settings.perplexity_api_key:
        return None

    response = httpx.post(
        EMBED_URL,
        headers={"Authorization": f"Bearer {settings.perplexity_api_key}"},
        json={"model": EMBED_MODEL, "input": text, "encoding_format": "base64_int8"},
        timeout=30.0,
    )
    response.raise_for_status()
    data = response.json()
    return _decode_int8(data["data"][0]["embedding"])

File: lib/embeddings.ts
import OpenAI from "openai";

/**
 * Real semantic embeddings via Perplexity's embeddings endpoint. Model IDs and
 * the base64_int8 encoding requirement verified directly against the live API
 * on 2026-06-27 — Perplexity rejects the standard OpenAI "float" encoding and
 * its own model names differ from OpenAI's (`pplx-embed-v1-0.6b`, not
 * `text-embedding-3-small`).
 */
const EMBED_MODEL = "pplx-embed-v1-0.6b";

let client: OpenAI | null = null;

function getClient(): OpenAI {
  if (client) return client;
  const apiKey = process.env.PERPLEXITY_API_KEY;
  if (!apiKey) {
    throw new Error("Missing PERPLEXITY_API_KEY in .env.development");
  }
  client = new OpenAI({ apiKey, baseURL: "https://api.perplexity.ai/v1" });
  return client;
}

/** Decodes a base64_int8-encoded embedding into a plain numeric vector. */
function decodeInt8(base64: string): number[] {
  const buf = Buffer.from(base64, "base64");
  return Array.from(new Int8Array(buf.buffer, buf.byteOffset, buf.length));
}

/** Embeds a batch of texts. Returns one vector per input, in order. */
export async function embed(texts: string[]): Promise<number[][]> {
  if (texts.length === 0) return [];
  const response = await getClient().embeddings.create({
    model: EMBED_MODEL,
    input: texts,
    encoding_format: "base64_int8",
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any);
  return response.data
    .sort((a, b) => a.index - b.index)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    .map((d) => decodeInt8((d as any).embedding as string));
}

/** Embeds a single text — convenience wrapper around `embed`. */
export async function embedOne(text: string): Promise<number[]> {
  const [vector] = await embed([text]);
  return vector;
}

/** Cosine similarity between two vectors of equal length. */
export function cosineSimilarity(a: number[], b: number[]): number {
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  if (normA === 0 || normB === 0) return 0;
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}
